xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,247 @@
+    {
+        return addAll(c);
+    }
+
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(this.decorated().get(index).getFilename());
+        if (removedAttachment != null) {
+            updatedMap();
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return true unless the attachment is not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        XWikiAttachment xwikiAttachment = (XWikiAttachment) attachment;
+        if (map.remove(xwikiAttachment.getFilename()) == null) {
+            return false;
+        }
+        updatedMap();
+        return true;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was previously matched to the same filename or null if no attachment was matched to
+     *         it
+     * @since 10.0RC1
+     */
+    public XWikiAttachment set(XWikiAttachment attachment)
+    {
+        XWikiAttachment put = map.put(attachment.getFilename(), attachment);
+        if (put != attachment) {
+            added(attachment);
+            updatedMap();
+        }
+        return put;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param index this parameter is not used but is needed to override the method
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was previously matched to the same filename or null if no attachment was matched to
+     *         it
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment set(int index, XWikiAttachment attachment)
+    {
+        return set(attachment);
+    }
+
+    /**
+     * @param filename the filename of the attachment to be returned.
+     * @return attachment with the given filename or null if not found.
+     * @since 10.0RC1
+     */
+    public XWikiAttachment getByFilename(String filename)
+    {
+        return map.get(filename);
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> c)
+    {
+        boolean changed = false;
+        for (XWikiAttachment x : (Collection<? extends XWikiAttachment>) c) {
+            if (map.remove(x.getFilename(), x)) {
+                changed = true;
+            }
+        }
+        if (changed) {
+            updatedMap();
+        }
+        return changed;
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c)
+    {
+        boolean changed = false;
+        Collection<XWikiAttachment> values = map.values();
+        for (XWikiAttachment x : values) {
+            if (!c.contains(x)) {
+                map.remove(x.getFilename(), x);
```

Comment:
Too quick, I doubt this builds :)

Reply:
@tmortagne looks fine on eclipse, do you mean a maven build?

Reply:
@tmortagne do I have the logic correct : 
while iterator has next value
if next is equal to x
remove using iterator and break

Reply:
Sorry I just forgot about the Map#get with two parameters...

Still I don't see why you don't use the other remove actually.

Reply:
> @tmortagne do I have the logic correct :
> while iterator has next value
> if next is equal to x
> remove using iterator and break

The logic is the same as what you have right now (don't see any break in tour current code). Simply instead of a foreach you should use an Iterator and call Iterator#remove() instead of Map#remove.

Reply:
@tmortagne you're right, it is unnecessary - I was making sure that the filename corresponds to the correct attachment but it will because we are iterating through the map.values()